{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Python Tools for Genome-wide Gene-level Association\n",
    "\n",
    "## Steven D. Shaw"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you would like to use this code, please feel free to do so, and give credit if it is due. I am happy to answer any questions regarding these tools, or, if you would like to work together on a project related to gene-level association, please contact me.\n",
    "\n",
    "Email: shawsd@umich.edu\n",
    "\n",
    "If you happen to notice any errors, please do let me know - we all make mistakes and we can always help eachother overcome and learn from our mistakes :)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### IMPORT PACKAGES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pandas_plink import read_plink1_bin\n",
    "from os.path import join\n",
    "from os import path\n",
    "from pandas_plink import get_data_folder\n",
    "from pandas_plink import read_plink\n",
    "import os\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "from datetime import datetime\n",
    "\n",
    "import statsmodels.api as sm\n",
    "import scipy.stats as stats\n",
    "from statsmodels.api import add_constant\n",
    "from sklearn import decomposition\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "import time\n",
    "\n",
    "from factor_analyzer import FactorAnalyzer\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import datetime"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DATA LOADING & MERGING\n",
    "\n",
    "## Genetic Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in genetic data\n",
    "    # QC PLINK vars: --mind 0.05 --geno 0.01 --maf 0.01\n",
    "    # Rename SNPs: --update-map SNP_kgpID2rsID.txt --update-name\n",
    "\n",
    "# Insert name of genetic data file - should be precleaned in PLINK\n",
    "gene_data_filename = ''\n",
    "    \n",
    "(bim, fam, bed) = read_plink(gene_data_filename, verbose=False)\n",
    "bim.set_index('chrom')\n",
    "fam.set_index('i')\n",
    "bim = bim.sort_values(by=['chrom', 'pos'])\n",
    "\n",
    "bim.to_csv(\"bim.csv\")\n",
    "\n",
    "########\n",
    "\n",
    "# Provide a genome gene set list (E.g., 'glist-hg19.txt')\n",
    "gene_set_filename = ''\n",
    "\n",
    "# List of genes in the human genome, and their location (chromosome, start, and end)\n",
    "hg = pd.read_csv(gene_set_filename, \n",
    "                 sep=' ', \n",
    "                 header=None, \n",
    "                 names=[\"chromosome\", \"Start\", \"End\", \"gene\"])\n",
    "\n",
    "# Add column that is current position just in case for joining afterwards\n",
    "hg['row_num'] = np.arange(len(hg))\n",
    "\n",
    "hg = hg.sort_values(by=['Start', 'End'])\n",
    "hg"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Survey Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Insert name of survey data file - this file should include: dependent variables, age, \n",
    "# population stratification information from genome-wide PCA (labelled EV1, EV2, etc.), and \n",
    "# IDs that match genetic data\n",
    "survey_data_filename = ''\n",
    "    \n",
    "DATA_survey = pd.read_csv(survey_data_filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MAP SNPs TO GENES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reorganize the genetic data (bim) to determine every gene that each SNP is located on\n",
    "# Use gene location information in hg file (e.g., glist-hg19.txt) to traverse through chromosomes, sequentially,\n",
    "# gene-by-gene and identify which SNPs are in each gene\n",
    "\n",
    "hg.set_index('chromosome')\n",
    "\n",
    "snp_to_gene_mapping = []\n",
    "for chrome_i in ['1', '2', '3', '4', '5', '6', '7', \n",
    "               '8', '9', '10', '11', '12', '13', '14', \n",
    "               '15', '16', '17', '18', '19', '20', '21', '22']:\n",
    "    now = datetime.now()\n",
    "    current_time = now.strftime(\"%H:%M:%S\")\n",
    "    print(\"Current Time =\", current_time)\n",
    "    print(\"chromosome: \", chrom_i)\n",
    "    chrom_x = bim.query(\"chrom=='{0}'\".format(chrom_i))\n",
    "    hg_x = hg.query(\"chromosome=='{0}'\".format(chrom_i))\n",
    "    start = hg_x['Start'].min()\n",
    "    end = hg_x['End'].max()\n",
    "    chrom_x = chrom_x[chrom_x[\"pos\"].between(start, end)]\n",
    "    length = len(chrom_x)\n",
    "\n",
    "    j = 0\n",
    "    \n",
    "    for ind, r in hg_x.iterrows():\n",
    "        while chrom_x.iloc[j, 3] < r['Start']:\n",
    "            j += 1\n",
    "            if j >= length:\n",
    "                break\n",
    "        if j >= length:\n",
    "                break\n",
    "        while chrom_x.iloc[j, 3] <= r['End']:\n",
    "            snp_to_gene_mapping.append([chrom_x.iloc[j, 0], chrom_x.iloc[j, 1], chrom_x.iloc[j, 2], \n",
    "                                        chrom_x.iloc[j, 3], chrom_x.iloc[j, 4], chrom_x.iloc[j, 5], chrom_x.iloc[j, 6], \n",
    "                                        r['Start'], r['End'], r['gene']])\n",
    "            j += 1\n",
    "            if j >= length:\n",
    "                break\n",
    "        if j >= length:\n",
    "                break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Organize traversal into dataframe\n",
    "bim_with_genes = pd.DataFrame(snp_to_gene_mapping, columns = ['chrom' , 'snp', 'cm', 'pos', \n",
    "                                                               'a0', 'a1', 'i','start', 'end', 'gene'])\n",
    "print(bim_with_genes)\n",
    "\n",
    "# Save this file, so you don't need to rerun above code on other computers\n",
    "# bim_with_genes.to_csv(\"bim_with_genes.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load in bim_with_genes for EigenSNP function\n",
    "#bim_with_genes = pd.read_csv('bim_with_genes.csv')\n",
    "#print(bim_with_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Required Functions\n",
    "\n",
    "#### PCA/ FA NOTES:\n",
    "\n",
    "Takes in a df where each row is a participant and each column is a SNP. Data input should be minor allele dosage.\n",
    "\n",
    "PCA is run once per gene in main function. Exploratory FA is run once to determine number of factors, followed by FA used for factor extraction.\n",
    "\n",
    "Output a df where each row is a participant and each column contains the eigenvectors from the PCA or FA factors.\n",
    "\n",
    "The number of PCs is determined by the cummulative variance required set by the 'limit' in the function (typically, greater than 90% variance explained). For FA, number of factors is calcaluate as factors with eigenvalues >1. \n",
    "\n",
    "#### REGRESION NOTES:\n",
    "\n",
    "Takes in two dfs:\n",
    "\n",
    "i_df: a df where rows are participants and columns are the independent variables.\n",
    "\n",
    "t_df: a df where rows are participants and column is the target dependent variable (i.e., BMI)\n",
    "\n",
    "NOTE: Both dataframes should have same participant order."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PCA Function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PC_eigen_vals(gene_df):\n",
    "    X = StandardScaler().fit_transform(gene_df)\n",
    "    limit = 0.9\n",
    "\n",
    "    #perform PCA\n",
    "    pca = decomposition.PCA(n_components=len(gene_df.columns))\n",
    "    pca.fit(X)\n",
    "    X = pca.transform(X)\n",
    "\n",
    "    #extract variance for each PC\n",
    "    var = pca.explained_variance_ratio_\n",
    "\n",
    "    # find num PCs\n",
    "    #np.cumsum(var)\n",
    "\n",
    "    # find num PCs\n",
    "    cum_var = 0\n",
    "    i = 0\n",
    "    while cum_var <= limit and i < len(var):\n",
    "        cum_var += var[i]\n",
    "        i += 1\n",
    "\n",
    "    # separate eigen vect columns based on limit\n",
    "    X_PCs = X[:,0:i]\n",
    "\n",
    "\n",
    "    # add to dictionary\n",
    "    PCs = {}\n",
    "    for i in range(0,X_PCs.shape[1]):\n",
    "        PCs[\"PC{0}\".format(i)] = pd.Series(X[:,i:i+1].transpose()[0])\n",
    "\n",
    "    # convert to df\n",
    "    PC_df = pd.DataFrame(PCs)\n",
    "\n",
    "    # Returns pd df\n",
    "    return PC_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FA Function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def FA_eigen_vals(gene_df):\n",
    "    X_norm = StandardScaler().fit_transform(gene_df)\n",
    "    limit = 1\n",
    "\n",
    "    #perform exploratory factor analysis to determine number of factors with eigenvalue over 1\n",
    "    efa = FactorAnalyzer(rotation=None, n_factors=len(gene_df.columns))\n",
    "    efa.fit(X_norm)\n",
    "    ev, v = efa.get_eigenvalues()\n",
    "    gene_fa = sum(ev > limit)\n",
    "\n",
    "    #Run factor analysis with number of factors from above EFA\n",
    "    gene_fa = FactorAnalyzer(rotation=\"promax\", n_factors=gene_fa)\n",
    "    gene_fa.fit(X_norm)\n",
    "\n",
    "    #extract eigen vector columns from FA\n",
    "    FA_eigen = gene_fa.fit_transform(X_norm)\n",
    "    #print(FA_eigen)\n",
    "\n",
    "    # add to dictionary\n",
    "    FAs = {}\n",
    "    for i in range(0,FA_eigen.shape[1]):\n",
    "        FAs[\"FA{0}\".format(i)] = pd.Series(FA_eigen[:,i:i+1].transpose()[0])\n",
    "\n",
    "    # convert to df\n",
    "    FA_df = pd.DataFrame(FAs)\n",
    "    return(FA_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Regression Function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def OLS_mod(i_df, t_df):\n",
    "    # set dep and ind variables\n",
    "    X = i_df[[i for i in i_df.columns]]\n",
    "    y = t_df[[i for i in t_df.columns]]\n",
    "\n",
    "    # add constant (intercepts)\n",
    "    X = sm.add_constant(X)\n",
    "\n",
    "    # least squares model\n",
    "    mod = sm.OLS(y, X).fit()\n",
    "    #print(mod.summary())\n",
    "\n",
    "    return mod"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GENE-LEVEL GENOME-WIDE ASSOCIATION FUNCTION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Provide a character list of the variables names for your dependent measures (from DATA_survey file)\n",
    "DV_list = ['']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Gene_results = []\n",
    "EigenSNPWAS_results = []\n",
    "FactanalSNPWAS_results = []\n",
    "\n",
    "# This loop takes each gene that we observe, obtains only the genetic dosage data that for that gene \n",
    "# from the main genetic dataset, then:\n",
    "    # conducts PCA and FA on SNPs within that gene, and retains eigenvectors from the number of PCs\n",
    "        # that explain >90% variance (i.e., eigenSNPs) or factors with eigenvalues >1\n",
    "    # merges the eigenSNPs with survey/non-genetic data for use with regression\n",
    "    # runs treatment and control regression models and compares these models to get the unique effect of the gene\n",
    "    # outputs the number of SNPs observed in the gene, the number of eigenSNPs, and the p-value\n",
    "\n",
    "Gene_results = pd.DataFrame(columns=['Gene', 'SNPs', 'ss_diff', 'f_stat','p_val'])\n",
    "EigenSNPWAS_results = pd.DataFrame(columns=['Gene', 'SNPs', 'EigenSNPs','PCA_ss_diff', 'PCA_f_stat','PCA_p_val'])\n",
    "FactanalSNPWAS_results = pd.DataFrame(columns=['Gene', 'SNPs', 'No_factors', 'FA_ss_diff', 'FA_f_stat','FA_p_val'])\n",
    "\n",
    "bim_with_genes.set_index('chrom')\n",
    "\n",
    "#Get a list of chromosomes in genetic dataset\n",
    "chrom_x = bim_with_genes['chrom'].unique()\n",
    "\n",
    "#Current time\n",
    "now = datetime.datetime.now()\n",
    "current_time = now.strftime(\"%H:%M:%S\")\n",
    "print(\"Current Time =\", current_time)\n",
    "\n",
    "#Start timer\n",
    "start_time = time.time()\n",
    "\n",
    "for j in chrom_x:\n",
    "\n",
    "    time_chk_chrom_loop = time.time()\n",
    "\n",
    "    print('Chromosome:', j)\n",
    "    chrom_df = bim_with_genes.query(\"chrom=={0}\".format(j))\n",
    "    #print(chrom_df)\n",
    "    #Create list of observed genes\n",
    "    chrom_df.set_index('gene')\n",
    "    genes_in_chrom_x = chrom_df['gene'].unique()\n",
    "    #print(genes_in_chrom_x)\n",
    "\n",
    "    #Chromosome counter so we can stop the loop, for testing\n",
    "    gene_counter = 0\n",
    "\n",
    "    #t1\n",
    "    time_chk_1 = time.time()\n",
    "\n",
    "    for i in genes_in_chrom_x:\n",
    "        #print(i)\n",
    "\n",
    "        time_chk_gene_loop = time.time()\n",
    "\n",
    "        #Specify how many genes to run - remove to run all\n",
    "    #    if gene_counter == 10:\n",
    "    #        break\n",
    "\n",
    "        #Display progress bar\n",
    "        if gene_counter in list(range(1, len(genes_in_chrom_x), 50)): #Change division size for different progress indicators\n",
    "            print('Gene:', gene_counter, 'of', len(genes_in_chrom_x))\n",
    "            elapsed_time = time.time() - start_time\n",
    "            print(\"Elapsed Time =\", time.strftime(\"%H:%M:%S\", time.gmtime(elapsed_time)))\n",
    "\n",
    "        # Get the SNP data from bed file\n",
    "        gene_df = chrom_df.query(\"gene=='{0}'\".format(i))\n",
    "        x = bed[gene_df.i.values, :].compute() #x contains only dosage data for SNPs in current gene\n",
    "\n",
    "        # transpose so that data are formatted correctly (columns are SNPs, rows are participants)\n",
    "        x_trans = x.transpose()\n",
    "        #print(x_trans)\n",
    "\n",
    "        # Get SNP names, create dataframe with SNP labels (as columns) for use in PCA\n",
    "        snps = gene_df['snp']\n",
    "        SNP_data = pd.DataFrame(x_trans, columns = snps)\n",
    "        #print('with NaN:', SNP_data)\n",
    "\n",
    "        #There are some strange columns of all NaNs, remove these first\n",
    "        SNP_data.dropna(axis=0, how='all', inplace=True)\n",
    "\n",
    "        #Replace NaNs with mean for each column\n",
    "        #SNP_data.fillna(SNP_data.mean(), axis=0, inplace=True)\n",
    "        SNP_data = SNP_data.apply(lambda x: x.fillna(x.mean()),axis=0)\n",
    "\n",
    "        #finally, if any NAs remain within a row, drop that row\n",
    "            #very few of these cases, but needed to keep function running (e.g., CSMD1 in chr8, has random missingness)\n",
    "        #SNP_data.dropna(axis=1, how='any', inplace=True)\n",
    "        #print(SNP_data.shape)\n",
    "        #print('without NaN:', SNP_data)\n",
    "\n",
    "        #Create a seperate dataset for PCA\n",
    "        SNP_data_for_PC = SNP_data.copy()\n",
    "\n",
    "        #Dataset for FA\n",
    "        #Need to remove any columns (SNPs) where standard deviation = 0\n",
    "        SNP_data_for_FA = SNP_data.drop(SNP_data.std()[SNP_data.std() == 0].index.values, axis=1)\n",
    "\n",
    "        #Count number of SNPs\n",
    "        SNP_count = len(SNP_data.columns)\n",
    "        SNP_count_PCA = len(SNP_data_for_PC.columns)\n",
    "        SNP_count_FA = len(SNP_data_for_FA.columns)\n",
    "        #print(i)\n",
    "        #print(SNP_count)\n",
    "\n",
    "    #t2\n",
    "        time_chk_2 = time.time()\n",
    "\n",
    "    ####################################################\n",
    "    ######RAW SNP DATA - NO PCA OR FA\n",
    "\n",
    "    ################################\n",
    "        skip_pca = 0\n",
    "        skip_fa = 0\n",
    "\n",
    "        if SNP_count_PCA > 1:\n",
    "            try:\n",
    "                PCA_og = PC_eigen_vals(SNP_data_for_PC)\n",
    "\n",
    "            except:\n",
    "                skip_pca = 1\n",
    "\n",
    "        if SNP_count_FA > 1:\n",
    "            try:\n",
    "                FA_og = FA_eigen_vals(SNP_data_for_FA)\n",
    "\n",
    "            except:\n",
    "                skip_fa = 1\n",
    "    ################################\n",
    "\n",
    "        #Traverse DV_list\n",
    "        for DV in DV_list:\n",
    "\n",
    "            if SNP_count == 0:\n",
    "                #EigenSNPWAS.append([i, 0, np.nan, np.nan])\n",
    "                Gene_results = Gene_results.append({'Gene': i,\n",
    "                                'SNPs': 0,\n",
    "                                'ss_diff': np.nan,\n",
    "                                'f_stat': np.nan,\n",
    "                                'p_val': np.nan,\n",
    "                                'DV': np.nan},\n",
    "                                ignore_index=True)\n",
    "\n",
    "            else:\n",
    "                try:\n",
    "\n",
    "                    if SNP_count == 1:\n",
    "                        Gene_reg = SNP_data.copy()\n",
    "                    else:\n",
    "                        Gene_reg = SNP_data.copy()\n",
    "\n",
    "                    # Take output of PCA - eigenSNPs - and merge with survey/non-genetic data\n",
    "                    #Get ID codes for gene data - use key values (i) from query with gene name\n",
    "                        #gene_df contains only the SNPs we are interested in\n",
    "\n",
    "                    #Append gene IDs to PCA output\n",
    "                    Gene_reg['IID'] = pd.to_numeric(fam['iid'], errors=\"coerce\")\n",
    "\n",
    "                    # new code - Add Sex variable from fam file\n",
    "                    Gene_reg['Sex'] = pd.to_numeric(fam['gender'], errors=\"coerce\")\n",
    "\n",
    "                    #merge eigenSNPs with dataset: DATA_survey\n",
    "                    Gene_reg = Gene_reg[Gene_reg['IID'].notna()]\n",
    "\n",
    "                    DATA_survey_DV = DATA_survey[[DV, 'IID', 'Age', 'EV1', 'EV2', 'EV3', 'EV4']]\n",
    "\n",
    "                    DATA_all_for_gene = pd.merge(DATA_survey_DV, Gene_reg, left_on='IID', right_on='IID', how='inner')\n",
    "\n",
    "                    #Drop the IID column, as it is no longer needed\n",
    "                    DATA_all_for_gene.drop('IID', axis=1, inplace=True)\n",
    "\n",
    "                    #Remove any rows with missing values\n",
    "                    DATA_all_for_gene.dropna(inplace=True)\n",
    "\n",
    "                    DATA_DV_only = pd.DataFrame(DATA_all_for_gene[DV])\n",
    "\n",
    "                    #Create treatment dataset\n",
    "                    DATA_treatment_model = DATA_all_for_gene.drop(DV, axis=1)\n",
    "\n",
    "                    #Create control dataset\n",
    "                    DATA_control_model = DATA_all_for_gene[['Age', 'Sex', 'EV1', 'EV2', 'EV3', 'EV4']]\n",
    "\n",
    "                    # TREATMENT Model\n",
    "                    treat_mod = OLS_mod(DATA_treatment_model, DATA_DV_only)     ## controls + EigenSNPs\n",
    "\n",
    "                    # CONTROL Model\n",
    "                    cont_mod = OLS_mod(DATA_control_model, DATA_DV_only)   ## controls only\n",
    "\n",
    "                    #Conduct ANOVA to compare TREATMENT and CONTROL models\n",
    "                    aov_table = sm.stats.anova_lm(cont_mod, treat_mod, typ=1)\n",
    "\n",
    "                    # Extract stats from model (F and p-val)\n",
    "                    gene_ss_diff = aov_table['ss_diff'][1]\n",
    "                    gene_fstat = aov_table['F'][1]\n",
    "                    gene_pval = aov_table['Pr(>F)'][1]\n",
    "                    #print('f:', gene_fstat, 'p:', gene_pval)\n",
    "\n",
    "                    Gene_results = Gene_results.append({'Gene': i,\n",
    "                                'SNPs': SNP_count_PCA,\n",
    "                                'ss_diff': gene_ss_diff,\n",
    "                                'f_stat': gene_fstat,\n",
    "                                'p_val': gene_pval,\n",
    "                                'DV': DV},\n",
    "                                ignore_index=True)\n",
    "\n",
    "                except:\n",
    "\n",
    "                    Gene_results = Gene_results.append({'Gene': i,\n",
    "                                        'SNPs': SNP_count_PCA,\n",
    "                                        'ss_diff': np.nan,\n",
    "                                        'f_stat': np.nan,\n",
    "                                        'p_val': np.nan,\n",
    "                                        'DV': DV},\n",
    "                                        ignore_index=True)\n",
    "\n",
    "\n",
    "\n",
    "    ####################################################\n",
    "    ####PCA\n",
    "\n",
    "            time_chk_3 = time.time()\n",
    "            #print('before PCA:', SNP_data_for_PC)\n",
    "\n",
    "            if SNP_count_PCA == 0:\n",
    "\n",
    "                EigenSNPWAS_results = EigenSNPWAS_results.append({'Gene': i,\n",
    "                                'SNPs': 0,\n",
    "                                'EigenSNPs': np.nan,\n",
    "                                'PCA_ss_diff': np.nan,\n",
    "                                'PCA_f_stat': np.nan,\n",
    "                                'PCA_p_val': np.nan,\n",
    "                                'DV': np.nan},\n",
    "                                ignore_index=True)\n",
    "            else:\n",
    "                try:\n",
    "\n",
    "                    if SNP_count_PCA == 1:\n",
    "                        PCA = SNP_data_for_PC.copy()\n",
    "                    else:\n",
    "                        #Conduct PCA - output contains dataframe with EigenSNPs (cut to >90% threshold)\n",
    "                        PCA = PCA_og.copy()\n",
    "                        if skip_pca == 1:\n",
    "                            raise Exception\n",
    "\n",
    "                    #Count number of EigenSNPs\n",
    "                    PC_Eigen_count = len(PCA.columns)\n",
    "\n",
    "                    # Take output of PCA - eigenSNPs - and merge with survey/non-genetic data\n",
    "                    #Get ID codes for gene data - use key values (i) from query with gene name\n",
    "                        #gene_df contains only the SNPs we are interested in\n",
    "\n",
    "                    #Append gene IDs to PCA output\n",
    "                    PCA['IID'] = pd.to_numeric(fam['iid'], errors=\"coerce\")\n",
    "                    # new code - Add Sex variable from fam file\n",
    "                    PCA['Sex'] = pd.to_numeric(fam['gender'], errors=\"coerce\")\n",
    "                    #print('fam iid_type:', PCA['IID'].dtypes)\n",
    "                    #print(PCA)\n",
    "\n",
    "                    #merge eigenSNPs with dataset: DATA_survey\n",
    "                    PCA = PCA[PCA['IID'].notna()]\n",
    "\n",
    "                    DATA_survey_DV = DATA_survey[[DV, 'IID', 'Age', 'EV1', 'EV2', 'EV3', 'EV4']]\n",
    "\n",
    "                    DATA_all_for_gene_PCA = pd.merge(DATA_survey_DV, PCA, left_on='IID', right_on='IID', how='inner')\n",
    "                    #print(list(DATA_all_for_gene.columns)) #View columns in dataset\n",
    "\n",
    "                    #Drop the IID column, as it is no longer needed\n",
    "                    DATA_all_for_gene_PCA.drop('IID', axis=1, inplace=True)\n",
    "                    #Remove any rows with missing values\n",
    "                    DATA_all_for_gene_PCA.dropna(inplace=True)\n",
    "\n",
    "                    DATA_DV_only_PCA = pd.DataFrame(DATA_all_for_gene_PCA[DV])\n",
    "\n",
    "                    #Create treatment dataset\n",
    "                    DATA_treatment_model_PCA = DATA_all_for_gene_PCA.drop(DV, axis=1)\n",
    "                    #print(DATA_treatment_model)\n",
    "\n",
    "                    #Create control dataset\n",
    "                    DATA_control_model_PCA = DATA_all_for_gene_PCA[['Age', 'Sex', 'EV1', 'EV2', 'EV3', 'EV4']]\n",
    "                    #print(DATA_control_model)\n",
    "\n",
    "                    # TREATMENT Model\n",
    "                    treat_mod_PCA = OLS_mod(DATA_treatment_model_PCA, DATA_DV_only_PCA)     ## controls + EigenSNPs\n",
    "\n",
    "                    #treat_mod.summary()\n",
    "\n",
    "                    # CONTROL Model\n",
    "                    cont_mod_PCA = OLS_mod(DATA_control_model_PCA, DATA_DV_only_PCA)\n",
    "                    #cont_mod.summary()\n",
    "\n",
    "                    #Conduct ANOVA to compare TREATMENT and CONTROL models\n",
    "                    aov_table_PCA = sm.stats.anova_lm(cont_mod_PCA, treat_mod_PCA, typ=1)\n",
    "                    #print(aov_table)\n",
    "\n",
    "                    # Extract stats from model (F and p-val)\n",
    "                    PCA_ss_diff = aov_table_PCA['ss_diff'][1]\n",
    "                    PCA_fstat = aov_table_PCA['F'][1]\n",
    "                    PCA_pval = aov_table_PCA['Pr(>F)'][1]\n",
    "\n",
    "                    #print('Gene:', i, 'SNPs:', SNP_count, 'EigenSNPs:', PC_Eigen_count, 'p_val:', gene_pval)\n",
    "\n",
    "                    EigenSNPWAS_results = EigenSNPWAS_results.append({'Gene': i,\n",
    "                                                                      'SNPs': SNP_count_PCA,\n",
    "                                                                      'EigenSNPs': PC_Eigen_count,\n",
    "                                                                      'PCA_ss_diff': PCA_ss_diff,\n",
    "                                                                      'PCA_f_stat': PCA_fstat,\n",
    "                                                                      'PCA_p_val': PCA_pval,\n",
    "                                                                      'DV': DV},\n",
    "                                                                      ignore_index=True)\n",
    "\n",
    "                except:\n",
    "\n",
    "                    EigenSNPWAS_results = EigenSNPWAS_results.append({'Gene': i,\n",
    "                                                      'SNPs': SNP_count_PCA,\n",
    "                                                      'EigenSNPs': np.nan,\n",
    "                                                      'PCA_ss_diff': np.nan,\n",
    "                                                      'PCA_f_stat': np.nan,\n",
    "                                                      'PCA_p_val': np.nan,\n",
    "                                                      'DV': DV},\n",
    "                                                      ignore_index=True)\n",
    "\n",
    "    ####################################################\n",
    "                ####FA\n",
    "\n",
    "            time_chk_4 = time.time()\n",
    "\n",
    "            if SNP_count_FA == 0:\n",
    "                FactanalSNPWAS_results = FactanalSNPWAS_results.append({'Gene': i,\n",
    "                    'SNPs': 0,\n",
    "                    'No_factors': np.nan,\n",
    "                    'FA_ss_diff': np.nan,\n",
    "                    'FA_f_stat': np.nan,\n",
    "                    'FA_p_val': np.nan,\n",
    "                    'DV': np.nan},\n",
    "                    ignore_index=True)\n",
    "\n",
    "            else:\n",
    "                try:\n",
    "\n",
    "                    if SNP_count_FA == 1:\n",
    "                        FA = SNP_data_for_FA.copy()\n",
    "\n",
    "                    else:\n",
    "                        #Conduct Factor Analysis\n",
    "                        FA = FA_og.copy()\n",
    "                        if skip_fa == 1:\n",
    "                            raise Exception\n",
    "\n",
    "                    #Count number of EigenSNPs\n",
    "                    FA_Eigen_count = len(FA.columns)\n",
    "                    #print(PC_Eigen_count)\n",
    "\n",
    "                    # Take output of PCA - eigenSNPs - and merge with survey/non-genetic data\n",
    "                    #Get ID codes for gene data - use key values (i) from query with gene name\n",
    "                        #gene_df contains only the SNPs we are interested in\n",
    "\n",
    "                    #Append gene IDs to PCA output\n",
    "                    FA['IID'] = pd.to_numeric(fam['iid'], errors=\"coerce\")\n",
    "                    # new code - Add Sex variable from fam file\n",
    "                    FA['Sex'] = pd.to_numeric(fam['gender'], errors=\"coerce\")\n",
    "                    #print('fam iid_type:', PCA['IID'].dtypes)\n",
    "                    #print(PCA)\n",
    "\n",
    "                    #merge eigenSNPs with dataset: DATA_survey\n",
    "                    FA = FA[FA['IID'].notna()]\n",
    "\n",
    "                    DATA_survey_DV = DATA_survey[[DV, 'IID', 'Age', 'EV1', 'EV2', 'EV3', 'EV4']]\n",
    "\n",
    "                    DATA_all_for_gene_FA = pd.merge(DATA_survey_DV, FA, left_on='IID', right_on='IID', how='inner')\n",
    "                    #print(list(DATA_all_for_gene.columns)) #View columns in dataset\n",
    "\n",
    "                    #Drop the IID column, as it is no longer needed\n",
    "                    DATA_all_for_gene_FA.drop('IID', axis=1, inplace=True)\n",
    "                    #Remove any rows with missing values\n",
    "                    DATA_all_for_gene_FA.dropna(inplace=True)\n",
    "\n",
    "                    DATA_DV_only_FA = pd.DataFrame(DATA_all_for_gene_FA[DV])\n",
    "\n",
    "                    #Create treatment dataset\n",
    "                    DATA_treatment_model_FA = DATA_all_for_gene_FA.drop(DV, axis=1)\n",
    "                    #print(DATA_treatment_model)\n",
    "\n",
    "                    #Create control dataset\n",
    "                    DATA_control_model_FA = DATA_all_for_gene_FA[['Age', 'Sex', 'EV1', 'EV2', 'EV3', 'EV4']]\n",
    "                    #print(DATA_control_model)\n",
    "\n",
    "                    # TREATMENT Model\n",
    "                    treat_mod_FA = OLS_mod(DATA_treatment_model_FA, DATA_DV_only_FA)     ## controls + EigenSNPs\n",
    "\n",
    "                    #treat_mod.summary()\n",
    "\n",
    "                    # CONTROL Model\n",
    "                    cont_mod_FA = OLS_mod(DATA_control_model_FA, DATA_DV_only_FA)\n",
    "                    #cont_mod.summary()\n",
    "\n",
    "                    #Conduct ANOVA to compare TREATMENT and CONTROL models\n",
    "                    aov_table_FA = sm.stats.anova_lm(cont_mod_FA, treat_mod_FA, typ=1)\n",
    "                    #print(aov_table)\n",
    "\n",
    "                    # Extract stats from model (F and p-val)\n",
    "                    FA_ss_diff = aov_table_FA['ss_diff'][1]\n",
    "                    FA_fstat = aov_table_FA['F'][1]\n",
    "                    FA_pval = aov_table_FA['Pr(>F)'][1]\n",
    "\n",
    "                    #print('Gene:', i, 'SNPs:', SNP_count, 'EigenSNPs:', PC_Eigen_count, 'p_val:', gene_pval)\n",
    "\n",
    "                    FactanalSNPWAS_results = FactanalSNPWAS_results.append({'Gene': i,\n",
    "                                                                            'SNPs': SNP_count_FA,\n",
    "                                                                            'No_factors': FA_Eigen_count,\n",
    "                                                                            'FA_ss_diff': FA_ss_diff,\n",
    "                                                                            'FA_f_stat': FA_fstat,\n",
    "                                                                            'FA_p_val': FA_pval,\n",
    "                                                                            'DV': DV},\n",
    "                                                                            ignore_index=True)\n",
    "\n",
    "                except:\n",
    "\n",
    "                    FactanalSNPWAS_results = FactanalSNPWAS_results.append({'Gene': i,\n",
    "                                                            'SNPs': SNP_count_FA,\n",
    "                                                            'No_factors': np.nan,\n",
    "                                                            'FA_ss_diff': np.nan,\n",
    "                                                            'FA_f_stat': np.nan,\n",
    "                                                            'FA_p_val': np.nan,\n",
    "                                                            'DV': DV},\n",
    "                                                            ignore_index=True)\n",
    "\n",
    "    #t4\n",
    "        #print('near end:', SNP_data_for_PC)\n",
    "        time_chk_end = time.time()\n",
    "        gene_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output datafiles for each gene-level representation\n",
    "Gene_results.to_csv()\n",
    "EigenSNPWAS_results.to_csv()\n",
    "FactanalSNPWAS_results.to_csv()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
